Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,368 @@
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
+    /// The default look group used to schedule async work
+    let loopGroup: EventLoopGroup
+
+    /// The configuration used to bootstrap new HTTP connections
+    let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property shouldn't be directly accessed, use the `ConnectionPool` subscript instead
+    var _connectionProviders: [Key: ConnectionProvider] = [:]
+
+    private let lock = Lock()
+
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.loopGroup = group
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    private enum RequiredAction {
+        case make(EventLoopPromise<ConnectionProvider>)
+        case existing(ConnectionProvider)
+    }
+
+    /// Determines what action should be taken regarding the the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurently from multiple threads
+    private func getAction(for key: Key, eventLoop: EventLoop) -> RequiredAction {
+        return self.lock.withLock {
+            if let provider = self._connectionProviders[key] {
+                return .existing(provider)
+            } else {
+                let promise = eventLoop.makePromise(of: ConnectionProvider.self)
+                promise.futureResult.whenComplete { result in
+                    switch result {
+                    case .success(let provider):
+                        self[key] = provider
+                    case .failure:
+                        self[key] = nil
+                    }
+                }
+                return .make(promise)
+            }
+        }
+    }
```

Comment:
not sure about the lock over the the entire block inc. the async operation. we could break it down and lock only the read and write parts. opinion?

Reply:
Thank you for noticing this, looking at this function and its callers I think there might be some threading issues I missed, I will also check how much we can reduce the span of lock closures.